contracts/tatum/Tatum721Provenance.sol

```diff
@@ -0,0 +1,233 @@
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override(ERC721, ERC721Enumerable) {
+        super._beforeTokenTransfer(from, to, tokenId);
+    }
+
+    function _burn(uint256 tokenId)
+        internal
+        virtual
+        override(ERC721, ERC721URIStorage)
+    {
+        return ERC721URIStorage._burn(tokenId);
+    }
+
+    function tokenCashbackValues(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (uint256[] memory)
+    {
+        return _cashbackValues[tokenId];
+    }
+
+    function tokenCashbackRecipients(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (address[] memory)
+    {
+        return _cashbackRecipients[tokenId];
+    }
+
+    function updateCashbackForAuthor(uint256 tokenId, uint256 cashbackValue)
+        public
+        returns (bool)
+    {
+        for (uint256 i = 0; i < _cashbackValues[tokenId].length; i++) {
+            if (_cashbackRecipients[tokenId][i] == _msgSender()) {
+                _cashbackValues[tokenId][i] = cashbackValue;
+                return true;
+            }
+        }
+        return true;
+    }
+
+    function burn(uint256 tokenId) public virtual {
+        //solhint-disable-next-line max-line-length
+        require(
+            _isApprovedOrOwner(_msgSender(), tokenId),
+            "ERC721Burnable: caller is not owner nor approved"
+        );
+        _burn(tokenId);
+    }
+    function uintToString(uint v) private pure returns (string memory) {
+        uint maxlength = 100;
+        bytes memory reversed = new bytes(maxlength);
+        uint i = 0;
+        while (v != 0) {
+            uint remainder = v % 10;
+            v = v / 10;
+            reversed[i++] = bytes1(uint8(48 + remainder));
+        }
+        bytes memory s = new bytes(i); // i + 1 is inefficient
+        for (uint j = 0; j < i; j++) {
+            s[j] = reversed[i - j - 1]; // to avoid the off-by-one error
+        }
+        string memory str = string(s);  // memory isn't implicitly convertible to storage
+        return str;
+    }
+    function safeTransfer(
+        address to,
+        uint256 tokenId,
+        string memory data,
+        uint256 value
+    ) public payable {
+        string memory dataValue=uintToString(value);
+        bytes memory bytesData=abi.encodePacked(data,"'''####'''",dataValue);
+        
+        if (_cashbackRecipients[tokenId].length != 0) {
+            // checking cashback addresses exists and sum of cashbacks
+            require(
+                _cashbackRecipients[tokenId].length != 0,
+                "CashbackToken should be of cashback type"
+            );
+            uint256 percentSum = 0;
+            for (uint256 i = 0; i < _cashbackValues[tokenId].length; i++) {
+                percentSum += _cashbackValues[tokenId][i];
+            }
+            uint256 sum = (percentSum * value) / 100;
```

Comment:
obtain value from data, not form the parameter

Reply:
value in percents should be in decimal, not floating - 100% = 1000000. So minimal percentage value is 0.0001% = 1.

Reply:
still not resolved - percentSum must be divided by 10000 not 100, because percentSum for 0.01% should be 1, because of the uint256 type

Reply:
this is still 100, not 10000!!